### PR TITLE
Add Image Recognition Benchmark with DistributedDataParallel

### PR DESCRIPTION
**Pull request by mmilenkoski**
This PR adds a benchmark for Image Recognition using DistributedDataParallel.

The changes from our original AllReduce benchmark are the following:
- The model is wrapped in DistributedDataParallel
- The optimizer is changed to the SGD optimizer provided by PyTorch, but the parameters are not changed.
- The default behaviour is changed to train on GPU because we are interested in comparing our models to `Multi-Process Single-GPU` mode of DDP.

I tried comparing this benchmark to the benchmark `cifar10-resnet20-all-reduce` using MPI and NCCL and got the following results on the light goal:

1. DDP with MPI:
70% Top 1 Validation Accuracy reached in 92.720 seconds, Compute: 82.6206567287445 seconds, Communication: 8.461318731307983 seconds
2. AllReduce with MPI:
70% Top 1 Validation Accuracy reached in 50.080 seconds, Compute: 26.204060792922974 seconds, Communication: 22.87956166267395 seconds
3. DDP with NCCL:
70% Top 1 Validation Accuracy reached in 26.203 seconds, Compute: 18.231900215148926 seconds, Communication: 7.181280612945557 seconds
4. AllReduce with NCCL:
70% Top 1 Validation Accuracy reached in 25.789 seconds, Compute: 15.225504159927368 seconds, Communication: 9.609550476074219 seconds

### CODE REVIEW

**Comment by martinjaggi** [comment]
nice, this is very encouraging that the times (and steps i guess) are very comparable.
can you add a short description if this implements the benchmark task 1a or deviates from it? so if looking at the steps mathematically, is it identical (up to slight differences in the optimizer compared to the one we [define in the task](https://mlbench.readthedocs.io/en/latest/benchmark-tasks.html#benchmark-task-descriptions)), or might there be other slight differences e.g. in dataloader etc?  and you also had only one GPU per VM, right?

other pros/cons compared to our code, in portability and compatibility with different backends etc?

we should maybe re-open the discussion as if we want to allow several GPUs per node, but can make scaling become less regular and representative

we can keep the issue here updated too (or move it to benchmark, sorry) https://github.com/mlbench/mlbench-core/issues/75

File: pytorch/imagerecognition/cifar10-resnet20-distributed-data-parallel/main.py [hunk 1]
```diff
@@ -0,0 +1,289 @@
+"""Training ResNet for CIFAR-10 dataset.
+
+This implements the 1a image recognition benchmark task,
```
Comment (martinjaggi) [inline pytorch/imagerecognition/cifar10-resnet20-distributed-data-parallel/main.py]: please comment on any differences or if it implements it precisely in all aspects of the task

File: pytorch/imagerecognition/cifar10-resnet20-distributed-data-parallel/main.py [hunk 2]
```diff
@@ -0,0 +1,289 @@
...
+)
+from mlbench_core.dataset.imagerecognition.pytorch import CIFAR10V1
+from mlbench_core.dataset.util.pytorch import partition_dataset_by_rank
+from mlbench_core.evaluation.goals import (
+    task1_time_to_accuracy_light_goal,
+    task1_time_to_accuracy_goal,
+)
+from mlbench_core.evaluation.pytorch.metrics import TopKAccuracy
+from mlbench_core.models.pytorch.resnet import ResNetCIFAR
+from mlbench_core.optim.pytorch.optim import CentralizedSGD
```
Comment (martinjaggi) [inline pytorch/imagerecognition/cifar10-resnet20-distributed-data-parallel/main.py]: not used?

File: pytorch/imagerecognition/cifar10-resnet20-distributed-data-parallel/main.py [hunk 3]
```diff
@@ -0,0 +1,289 @@
...
+            dataset_dir,
+            ckpt_run_dir,
+            output_dir,
+            validation_only,
+            use_cuda=gpu,
+            light_target=light_target,
+        )
+
+
+if __name__ == "__main__":
```
Comment (martinjaggi) [inline pytorch/imagerecognition/cifar10-resnet20-distributed-data-parallel/main.py:233]: is this needed in each main.py for each implementation or should this be provided somewhere by mlbench core?

File: pytorch/imagerecognition/cifar10-resnet20-distributed-data-parallel/main.py (hunk 1 above)
Comment (mmilenkoski) [reply pytorch/imagerecognition/cifar10-resnet20-distributed-data-parallel/main.py]: It implements all of the same steps as the benchmark `cifar10-resnet20-all-reduce`. The only difference is that we wrap the model (which has the exact same definition as previously) in a DDP object and we use the SGD optimizer provided by PyTorch. DDP handles the allreduce step itself, so we don't need to use our own implementation of CentralisezSGD. More details on the implementation of DDP can be found [here](https://pytorch.org/docs/stable/notes/ddp.html). The only other change here is that we set the default behaviour to use GPU. You can see the difference in the `main.py` between the benchmark `cifar10-resnet20-all-reduce` and this one on this [link](http://www.mergely.com/k2EPXqI4/). They also use the same Dockerfile and requirements.

File: pytorch/imagerecognition/cifar10-resnet20-distributed-data-parallel/main.py (hunk 3 above)
Comment (mmilenkoski) [reply pytorch/imagerecognition/cifar10-resnet20-distributed-data-parallel/main.py:233]: I am not sure what you mean. We need to parse the arguments inside the benchmark.

File: pytorch/imagerecognition/cifar10-resnet20-distributed-data-parallel/main.py (hunk 1 above)
Comment (martinjaggi) [reply pytorch/imagerecognition/cifar10-resnet20-distributed-data-parallel/main.py]: perfect. want to add that info to the .py comment section?

Comment (mmilenkoski) [reply pytorch/imagerecognition/cifar10-resnet20-distributed-data-parallel/main.py]: you mean inside the docstring here?

**Comment by martinjaggi** [comment]
yes

**Comment by martinjaggi** [comment]
@Panaetius  this is not a reference implementation, but still some implementation of task 1a. where do we put it? (docu, image, code?) this could also serve as an example of a user-provided implementation, as to what people need to do if they want to bring their own algorithm into mlbench

**Review by martinjaggi — APPROVED** [review]

File: pytorch/imagerecognition/cifar10-resnet20-distributed-data-parallel/main.py (hunk 3 above)
Comment (Panaetius) [reply pytorch/imagerecognition/cifar10-resnet20-distributed-data-parallel/main.py:233]: Could be looked into in https://github.com/mlbench/mlbench-core/issues/85

**Comment by mmilenkoski** [comment]
Can I merge this, or should we move it to another place?

**Comment by martinjaggi** [comment]
yes you can merge it. but please clarify the description/docu that this is not the reference implementation for the task, but it is **a** impl for the same task. maybe we can add 'alternative-implementation' to the filename or a folder?

**Comment by martinjaggi** [comment]
also did you check with @ehoelzl if this one was also affected by NCCL stalling issue or not?
